pennylane/io/qualtran_io.py

```diff
@@ -418,3 +584,358 @@ def compute_matrix(*params, **hyperparams):  # pylint: disable=no-self-argument
+        if not qualtran:
+            raise ImportError(self._error_message)
+
+        if not isinstance(op, Operator) and not isinstance(op, QNode):
+            raise TypeError(f"Input must be either an instance of {Operator} or {QNode}.")
+
+        self.op = op
+        self.map_ops = map_ops
+        self._kwargs = kwargs
+        super().__init__()
+
+    @cached_property
+    def signature(self) -> "qt.Signature":
+        """Compute and return Qualtran signature for given op or QNode."""
+        if isinstance(self.op, QNode):
+            self.op.name = "QNode"
+            num_wires = len(construct_tape(self.op)(**self._kwargs).wires)
+        else:
+            num_wires = len(self.op.wires)
+        return qt.Signature([qt.Register("qubits", qt.QBit(), shape=num_wires)])
+
+    def decompose_bloq(self):  # pylint:disable=too-many-branches
+        """Decompose the bloq using the op's decomposition or the tape of the QNode"""
+        try:
+            if isinstance(self.op, QNode):
+                tape = construct_tape(self.op)(**self._kwargs)
+                ops = tape.circuit
+                all_wires = list(tape.wires)
+            else:
+                ops = self.op.decomposition()
+                all_wires = list(self.op.wires)
+
+            signature = self.signature
+            in_quregs = out_quregs = {"qubits": np.array(all_wires).reshape(len(all_wires), 1)}
+
+            in_key = list(in_quregs.keys())[0]
+            out_key = list(out_quregs.keys())[0]
+
+            in_quregs = {
+                in_key: np.apply_along_axis(
+                    _QReg, -1, in_quregs[in_key], signature.get_left(in_key).dtype
+                )
+            }
+            out_quregs = {
+                out_key: np.apply_along_axis(
+                    _QReg, -1, out_quregs[out_key], signature.get_right(out_key).dtype
+                )
+            }
+
+            bb, initial_soqs = qt.BloqBuilder.from_signature(signature, add_registers_allowed=False)
+
+            # `signature.lefts()` can be thought of as input qubits. For our purposes LEFT and
+            # RIGHT signatures will (almost) always match since there are no allocated & freed
+            # qubits. Here, qreg_to_qvar is a map between a register and a Soquet. This serves
+            # as the foundation to wire up the rest of the bloqs.
+            qreg_to_qvar = {}
+            for reg in signature.lefts():
+                assert reg.name in in_quregs
+                soqs = initial_soqs[reg.name]
+                assert in_quregs[reg.name].shape == soqs.shape
+                qreg_to_qvar |= zip(in_quregs[reg.name].flatten(), soqs.flatten())
+
+            # 2. Add each operation to the composite Bloq.
+            for op in ops:
+                bloq = _map_to_bloq(op, map_ops=self.map_ops)
+                if bloq is None:
+                    continue
+
+                if bloq.signature == qt.Signature([]):
```

Comment:
What does this do? What would an empty signature mean?

Reply:
It is an extreme edge case. Almost all bloqs have a signature but a bloq like `Identity(0)` would have an empty signature. If they don't have a signature we just need to "add" it to the bloq but we don't have to wire it up. So Identity(0) won't raise an error it will just be passed along. It's pretty much only relevant if someone decided to map a bloq to a bloq that had an empty signature (for what reason? don't know, but it should be handled).